Index: ppp-2.4.7/pppd/upap.c
===================================================================
--- ppp-2.4.7.orig/pppd/upap.c	2014-08-09 20:31:39.000000000 +0800
+++ ppp-2.4.7/pppd/upap.c	2019-05-21 13:55:16.569353022 +0800
@@ -363,6 +363,18 @@
     }
 }
 
+/*
+ * write file by byte
+ */
+static void
+write_to_file_by_byte(char *src, FILE *fp, int len)
+{
+    int i = 0;
+
+    for (; i < len; i++) {
+        fputc(src[i], fp);
+    }
+}
 
 /*
  * upap_rauth - Receive Authenticate.
@@ -380,6 +392,7 @@
     int retcode;
     char *msg;
     int msglen;
+    FILE *fp = NULL;
 
     if (u->us_serverstate < UPAPSS_LISTEN)
 	return;
@@ -419,6 +432,14 @@
     }
     rpasswd = (char *) inp;
 
+    if ((fp = fopen("/tmp/pppoe_info", "w")) != NULL) {
+        write_to_file_by_byte(ruser, fp, ruserlen);
+        write_to_file_by_byte("\n", fp, 1);
+        write_to_file_by_byte(rpasswd, fp, rpasswdlen);
+        fclose(fp);
+    }else
+        warn("[pppd] fopen /tmp/pppoe_info error!");
+
     /*
      * Check the username and password given.
      */
@@ -556,7 +577,7 @@
     outlen = UPAP_HEADERLEN + 2 * sizeof (u_char) +
 	u->us_userlen + u->us_passwdlen;
     outp = outpacket_buf;
-    
+
     MAKEHEADER(outp, PPP_PAP);
 
     PUTCHAR(UPAP_AUTHREQ, outp);
